[RDY] term: use an argument vector for termopen().

splinterofchaos opened the pull request:
Old behaviour: termopen('cmd') would run `&shell &shcf "cmd"`, which
caused the functional tests to fail on some systems due to the process
not "owning" the terminal. Also, it is inconsistent with jobstart().

Modify termopen() so that &shell is not invoked, but maintain the old
behaviour with :terminal. Factor the common code for building the
argument vector from jobstart() and modify the functional tests to call
termopen() instead of :terminal (fixes #2354).

---

Still [WIP] because I have yet to implement a correct naming scheme for the buffer (https://github.com/neovim/neovim/issues/2354#issuecomment-92532588), though I'm considering an alternative: add a `'name'` field to the options dictionary and set it to `{cmd}` from `ex_terminal`. Could be useful in general.

Also, I added a few tests for `:terminal`, since its behaviour now differs from `termopen()`, but I couldn't avoid using the Unix-specific `sh` without rewriting `tty-test.c` to not propagate the problem from #2354, or make another test shell. 

Will work on this more tomorrow.

Commit:
term: use an argument vector for termopen().

Old behaviour: termopen('cmd') would run `&shell &shcf "cmd"`, which
caused the functional tests to fail on some systems due to the process
not "owning" the terminal. Also, it is inconsistent with jobstart().

Modify termopen() so that &shell is not invoked, but maintain the old
behaviour with :terminal. Factor the common code for building the
argument vector from jobstart() and modify the functional tests to call
termopen() instead of :terminal (fixes #2354).

Also:
 * Add a 'name' option for termopen() so that `:terminal {cmd}` produces
   a buffer named "term//{cwd}/{cmd}" and termopen() users can customize
   the name.
 * Update the documentation.
 * Add functional tests for `:terminal` sinse its behaviour now differs
   from termopen(). Add "test/functional/fixtures/shell-test.c" and move
   "test/functional/job/tty-test.c" there, too.

Helped-by: Justin M. Keyes <@justinmk>

Pyrohh commented on src/nvim/eval.c:
```diff
@@ -10779,47 +10779,62 @@ static void f_jobresize(typval_T *argvars, typval_T *rettv)
 
-// "jobstart()" function
-static void f_jobstart(typval_T *argvars, typval_T *rettv)
+static char** list_to_argv(list_T *args)
```
pointer alignment (I like the way you're doing it, but it's not project style)

splinterofchaos replied:
:+1:

ZyX-I commented:
> Modify termopen() so that &shell is not invoked, but maintain the old
> behaviour with :terminal. Factor the common code for building the
> argument vector from jobstart() and modify the functional tests to call
> termopen() instead of :terminal (fixes #2354).

Why not just use `&shell` if `termopen` receives a string and new code when it receives a list? And, if possible, make the code that process this first argument reusable by `system` and `systemlist`.

splinterofchaos commented:
> Why not just use &shell if termopen receives a string and new code when it receives a list?

Wouldn't be difficult, but what's the motivation, backwards compatibility? If anything, I think it'd be more consistent to make `termopen("do a cmd")` equivalent to `termopen(['do', 'a', 'cmd'])` and maybe add an option to `{opts}` (either a boolean like `'use_shell'`, or `'shell'` and `'shellcmdflag'`).

> And, if possible, make the code that process this first argument reusable by system and systemlist

I do like that idea.

ZyX-I commented:
> Wouldn't be difficult, but what's the motivation, backwards compatibility?

It and consistency: you can’t remove `system(string)` due to backwards compatibility. And `termopen` also accepts shell command.

Also sometimes you *do* want to make command be processed by a shell (e.g. `call termopen('vim *.py')`) or do not have any special characters expect for space and raw string is more convenient to type.

// Side note: once we have a linter for VimL with these and changes to `system*` `termopen(string)`, `system(string)` and `systemlist(string)` should yield a warning.

bfredl commented:
I also think keeping `termopen(string)` as convenient alias, and then I think it should be supported for `jobopen(string, ...)` as well, as these function already have similar api (and purpose) it would be confusing to support it in one but not the other (also they could share `val_to_argv` similar to `list_to_argv` already in this PR, and perhaps even with `system` as @ZyX-I  noted).

splinterofchaos commented:
I'm starting to like these ideas, but right now I'm just working on removing the dependence of `sh` for running the tests. Maybe it's not a big deal until we get nvim on windows working, but if we get it out of the way right now it'll be less work later.

Also, the goal is to resolve #2354 so @Pyrohh, I, and whoever else can run functional tests without hacking into `test/functional/tty-test.c`, so I want to delay any new features that could delay the merging of this PR. Though, if we determine that `termopen()` is too widely used to *temporarily not* have backwards compatibility, then I'll have to fix that immediately anyway.

bfredl commented:
Sounds reasonable, no need to do it all in this PR. (Unless for the `termopen()` part, I don't know if it is relied upon)

splinterofchaos commented:
Silly me: spending all day writing a very simple shell to execute programs through libuv, just for testing `:terminal`, when all I had to so was print a prompt, the command, and exit.

Marking RFC.

Commit:
unify jobstart, termopen, and system interfaces

For any of these functions, if {cmd} is a string, execute
"&shell &shellcmdflag '{cmd}'", or simply {cmd} if it's a list.

In termopen(), if the 'name' option is not supplied, try to guess using
'{cmd}' (string) or {cmd}[0] (list).  Simplify ex_terminal to use the
string form of termopen().

termopen: get name from argument

Convert list_to_argv to tv_to_argv.

Helped-by: Björn Linse <@bfredl>
Helped-by: oni-link <knil.ino@gmail.com>
Helped-by: Thiago de Arruda <@tarruda>

splinterofchaos commented:
I've implemented the suggested change, but as I said, I don't want to delay merging this because of them. While reviewing, please let me know if the commits, minus the last, look good so that if I need to work on it more, but everything else is fine, I can remove the commit, merge, and create a new PR.

ZyX-I commented on runtime/doc/eval.txt:
```diff
@@ -6140,7 +6143,9 @@ synstack({lnum}, {col})					*synstack()*
-		Get the output of the shell command {expr} as a string.  See
+		Get the output of the shell command {expr} as a string. If
+		{expr} is a string, 'shell' and 'shellcmdflag' will be used,
+		if {expr} is a string, it will be run directly.  See
```
Here should be “`if {expr} is a |List|`”, not “if `{expr}` is a string”.

splinterofchaos replied:
:+1:

ZyX-I commented on runtime/doc/eval.txt:
```diff
@@ -6287,15 +6289,19 @@ tempname()					*tempname()* *temp-file-name*
+termopen({cmd}[, {opts}])			{Nvim} *termopen()*
+		Spawns {cmd} in a new pseudo-terminal session connected
+		to the current buffer. If {cmd} is a string, it will be
+		executed as "'shell' 'shellcmdflag' {cmd}".  This function
```
Missing “if {cmd} is a |List|” case.

splinterofchaos replied:
:+1:

ZyX-I commented on src/nvim/eval.c:
```diff
@@ -10779,47 +10779,71 @@ static void f_jobresize(typval_T *argvars, typval_T *rettv)
-      || (argvars[1].v_type != VAR_DICT && argvars[1].v_type != VAR_UNKNOWN)) {
-    // Wrong argument types
+  if (cmd->v_type != VAR_LIST) {
     EMSG(_(e_invarg));
```
I would highly suggest to *never* use `e_invarg`. It is not telling the user what is the error: should be something `Exxx: Expected List or String`.

splinterofchaos replied:
I moved the error checking down so that this gets caught by `get_tv_string`.

ZyX-I commented on src/nvim/eval.c:
```diff
@@ -10779,47 +10779,71 @@ static void f_jobresize(typval_T *argvars, typval_T *rettv)
+  list_T *args = cmd->vval.v_list;
   for (listitem_T *arg = args->lv_first; arg != NULL; arg = arg->li_next) {
     if (arg->li_tv.v_type != VAR_STRING) {
       EMSG(_(e_invarg));
```
Same.

splinterofchaos replied:
:+1:

ZyX-I commented on src/nvim/eval.c:
```diff
@@ -10779,47 +10779,71 @@ static void f_jobresize(typval_T *argvars, typval_T *rettv)
 
+  if (argvars[1].v_type != VAR_DICT && argvars[1].v_type != VAR_UNKNOWN) {
+    // Wrong argument types
+    EMSG(_(e_invarg));
```
Same.

splinterofchaos replied:
Now:     EMSG(_(e_invarg2), "expected dictionary");`

ZyX-I commented on test/functional/terminal/cursor_spec.lua:
```diff
@@ -144,7 +144,7 @@ describe('cursor with customized highlighting', function()
     })
     screen:attach(false)
-    execute('term "' ..nvim_dir.. '/tty-test"')
+    execute('call termopen(["' ..nvim_dir.. '/tty-test"]) | startinsert')
```
It is rather strange that `..nvim_dir..` is surrounded by spaces, but there are no spaces between both `..` and `nvim_dir`.

ZyX-I replied:
And same thing in other tests. Is this intentional?

splinterofchaos replied:
Thanks; a copy and paste error.

splinterofchaos commented:
@jszakmeister, Is there a timeout on QuickBuild? I think this one's stuck in an infinite loop. Been running for over two hours. http://neovim-qb.szakmeister.net/build/564/overview

splinterofchaos commented:
Rebased and squashed related commits together.

bfredl commented on src/nvim/os/shell.c:
```diff
@@ -166,17 +168,16 @@ int os_call_shell(char_u *cmd, ShellOpts opts, char_u *extra_args)
 
-static int shell(const char *cmd,
-                 const char *extra_args,
+static int shell(char **argv,
```
I don't think this function should be called `shell` as it doesn't necessarily use the shell anymore. perhaps `call_argv` or `call_system` or something like that.

splinterofchaos replied:
:+1: -- I vote for `do_os_system()`.

splinterofchaos commented:
Odd failure locally:
> ./test/functional/helpers.lua:135: attempt to perform arithmetic on local 'written' (a nil value)

From `request()` returning `nil` in `nvim_feed()`:
```lua
local function nvim_feed(input)
  while #input > 0 do
    local written = request('vim_input', input)
    input = input:sub(written + 1)
  end
end
```
Not sure why this would happen, but curious if travis will show the same.

splinterofchaos commented:
OK, a memory error caused it. `do_os_system()` (was `shell()`) now copies `argv[0]` to a temporary buffer for error reporting. Inefficient, but nothing compared to spawning a new program.

bfredl commented on runtime/doc/eval.txt:
```diff
@@ -4031,9 +4031,11 @@ jobsend({job}, {data})					{Nvim} *jobsend()*
-		dictionary with any of the following keys:
+jobstart({cmd}[, {opts}])				{Nvim} *jobstart()*
+		Spawns {cmd} as a job.  If {cmd} is a |String|, it will be ran
+		via "'shell' 'shellcmdflag' {cmd}", or directly if a |List|.
```
Perhaps remove the outer quotes here, it could be interpreted as we're concatenating a string here, but we're really building up an argument list.

splinterofchaos replied:
:+1:

bfredl commented on runtime/doc/eval.txt:
```diff
@@ -6299,15 +6301,20 @@ tempname()					*tempname()* *temp-file-name*
+termopen({cmd}[, {opts}])			{Nvim} *termopen()*
+		Spawns {cmd} in a new pseudo-terminal session connected
+		to the current buffer. If {cmd} is a |String|, it will be
+		executed as "'shell' 'shellcmdflag' {cmd}", otherwise if {cmd}
```
similarly here. Perhaps explicitly say something like "{cmd} is interpreted exactly like in |jobstart| " ?

splinterofchaos replied:
:+1:

bfredl commented on runtime/doc/eval.txt:
```diff
@@ -6152,7 +6155,9 @@ synstack({lnum}, {col})					*synstack()*
 		valid positions.
 
 system({expr} [, {input}])				*system()* *E677*
```
perhaps should change first argument to `{cmd}` as it works the same as the `{cmd}` in `jobstart`. Even if the explanation is repeated, I think it also should point out that it works the same as `jobstart`. This way, it is easier to reason if it is a "safe" refactor to change between the functions.

splinterofchaos replied:
:+1: -- forgot about `system()`'s docs.

bfredl commented on src/nvim/os/shell.c:
```diff
@@ -119,14 +119,14 @@ int os_call_shell(char_u *cmd, ShellOpts opts, char_u *extra_args)
-                     emsg_silent,
-                     forward_output);
+
+  int status = do_os_system(shell_build_argv((char *)cmd, (char *)extra_args),
```
Not really related to this PR, but if I haven't missed something it looks like `extra_args` is only ever used in `mch_expand_wildcards`. Perhaps this function could be refactored  to manipulate the argv list directly, instead of complicating the entire `call_shell` `os_call_shell` `shell_build_argv` chain. (later, no need to do it in this PR).

bfredl commented:
LGTM, just some comments on the documentation.

A further change could be to bring `rpcstart` into the fold, I see no point why it should have a different `rpcstart(arg0, [arg1, ...])` form.

splinterofchaos commented:
I opened issue #2463 to discuss whether `rpcstart`'s interface can change.

bfredl commented:
Great work @splinterofchaos ! Perhaps this PR should be merged as-is as it currently has no backward-incompatible changes (right?) and already is very useful, and then a PR for #2463 (if no objections to that)

tarruda commented:
LGTM, great work @splinterofchaos

oni-link commented on src/nvim/eval.c:
```diff
@@ -10787,47 +10787,74 @@ static void f_jobresize(typval_T *argvars, typval_T *rettv)
+      while (i-- > 0) {
+        xfree(argv[i]);
+      }
+      return NULL;
```
`argv` is not freed.

splinterofchaos replied:
:+1:

oni-link commented on runtime/doc/eval.txt:
```diff
@@ -4031,9 +4031,12 @@ jobsend({job}, {data})					{Nvim} *jobsend()*
-		dictionary with any of the following keys:
+jobstart({cmd}[, {opts}])				{Nvim} *jobstart()*
+		Spawns {cmd} as a job.  If {cmd} is a |List|, it will be run
+		director. If {cmd} is a |String|, it will be equivalent to >
```
What does `it will be run director` mean?

tarruda replied:
`it will be run directly`(passed to the operating system process spawning function)

splinterofchaos replied:
:+1:  -- typo

oni-link commented on src/nvim/eval.c:
```diff
@@ -10787,47 +10787,74 @@ static void f_jobresize(typval_T *argvars, typval_T *rettv)
     return;
   }
 
+  char **argv = tv_to_argv(&argvars[0]);
```
Just wondering, don't we check arguments in order, so that the first error message we see is for the first wrong argument?

tarruda replied:
Maybe refactor `tv_to_argv` so it receives a `char **` out parameter with the command name?

splinterofchaos replied:
That simplified `termopen()`, but did you have something in mind for `jobstart()`?

oni-link commented on src/nvim/eval.c:
```diff
@@ -15195,10 +15222,18 @@ static void f_termopen(typval_T *argvars, typval_T *rettv)
+  char *name = job_opts ?
+    (char *)get_dict_string(job_opts, (char_u *)"name", false) : NULL;
+  if (!name) {
+    typval_T *name_tv = (argvars[0].v_type == VAR_LIST) ?
```
Can't we use `argv` to get the name?

oni-link commented on src/nvim/eval.c:
```diff
@@ -10787,44 +10787,79 @@ static void f_jobresize(typval_T *argvars, typval_T *rettv)
+    char *a = (char *)get_tv_string_chk(&arg->li_tv);
+    if (a == NULL) {
+      // Did emsg in get_tv_string; just deallocate argv.
+      argv[i] = NULL;
```
All entries in `argv` started to be `NULL`, so this could be removed.

splinterofchaos replied:
:+1:

Pyrohh commented on runtime/doc/eval.txt:
```diff
@@ -4052,7 +4055,8 @@ jobstart({argv}[, {opts}])			{Nvim} *jobstart()*
 		  - 0 when the job table is full or on invalid arguments
-		  - -1 when {argv}[0] is not executable
+		  - -1 when {cmd}[0] is not executable. (Will never fail if
+		    {cmd} is a string unless 'shell' is not executable.)
```
It seems odd to have a sentence inside parentheses, maybe remove the parens? also there should be two spaces after period.

splinterofchaos replied:
:+1:

Pyrohh commented on runtime/doc/eval.txt:
```diff
@@ -6151,9 +6155,10 @@ synstack({lnum}, {col})					*synstack()*
+system({cmd} [, {input}])				*system()* *E677*
+		Get the output of the shell command {cmd} as a string.  {cmd}
+		will be run the same as in |jobstart()|.  See |systemlist()|
+		to get the output as a List.
```
Maybe `List` should be `|List|`, like `|String|` above.

splinterofchaos replied:
:+1:

Pyrohh commented on runtime/doc/eval.txt:
```diff
@@ -6182,11 +6187,8 @@ system({expr} [, {input}])				*system()* *E677*
-		For Unix braces are put around {expr} to allow for
-		concatenated commands.
+		The command executed is constructed using several options when
+		{expr} is a string: 'shell' 'shellcmdflag' {expr}
```
should be `{cmd}`

Pyrohh replied:
same with line 6175

splinterofchaos replied:
:+1:

Pyrohh commented on runtime/doc/eval.txt:
```diff
@@ -6299,15 +6301,18 @@ tempname()					*tempname()* *temp-file-name*
-		|jobstart()|.
+termopen({cmd}[, {opts}])			{Nvim} *termopen()*
+		Spawns {cmd} in a new pseudo-terminal session connected
+		to the current buffer. {cmd} is the same as the one passed to
```
two spaces after period

splinterofchaos replied:
:+1:

Pyrohh commented on runtime/doc/eval.txt:
```diff
@@ -6299,15 +6301,18 @@ tempname()					*tempname()* *temp-file-name*
+		but the `pty`, `width`, `height`, and `TERM` fields are
+		ignored: `height`/`width` are taken from the current window
+		and `$TERM` is set to "xterm-256color".  An additional option,
+		`name`, can be used to give the terminal a specific name.
```
Maybe ```... "xterm-256color". If `name` is specified, the teminal's name is set to `name`.```

splinterofchaos replied:
I feel that `'name'` must be noted as being unique to `termopen()`, I changed it to:
> "xterm-256color", and it may have a
		`name` field.  `name`, if present, sets the buffer's name to
		"term://{cwd}/{pid}:{name}".
		Returns the same values as |jobstart()|.

Pyrohh commented on runtime/doc/various.txt:
```diff
@@ -226,10 +226,17 @@ g8			Print the hex values of the bytes used in the
+			      :startinsert
 <
+                        If no {command} is given, 'shellcmdflag' will not be
+                        sent to |termopen()|.
```
should be tabs, not spaces

splinterofchaos replied:
:+1:

Pyrohh commented on runtime/doc/various.txt:
```diff
@@ -226,10 +226,17 @@ g8			Print the hex values of the bytes used in the
 :term[inal][!] {cmd}	Spawns {command} using the current value of 'shell'
-			in a new terminal buffer.  This is equivalent to: >
+                        and 'shellcmdflag' in a new terminal buffer.  This is
+                        equivalent to: >
```
should be tabs, not spaces

splinterofchaos replied:
:+1:

Pyrohh commented on runtime/doc/eval.txt:
```diff
@@ -4031,9 +4031,12 @@ jobsend({job}, {data})					{Nvim} *jobsend()*
-		dictionary with any of the following keys:
+jobstart({cmd}[, {opts}])				{Nvim} *jobstart()*
+		Spawns {cmd} as a job.  If {cmd} is a |List|, it will be run
+		directly.  If {cmd} is a |String|, it will be equivalent to >
```
The actual help option is `|string|`, so this should either be changed to that, or `*string*` in `eval.txt` should have `*String*` added next to it for consistency with `*List*` and `*list*`.

Pyrohh commented on src/nvim/eval.c:
```diff
@@ -10778,44 +10778,78 @@ static void f_jobresize(typval_T *argvars, typval_T *rettv)
-  rettv->vval.v_number = 0;
+  if (cmd_tv->v_type == VAR_STRING) {
+    char *cmd_str = (char *)get_tv_string(cmd_tv);
+    if (cmd) {
```
Explicit comparisons with null are used below, so it's probably a good idea to be consistent about it.

splinterofchaos replied:
:+1: -- This function has many `NULL` comparisons. I decided to just change the explicit tests to implicite.

Pyrohh commented on src/nvim/eval.c:
```diff
@@ -10778,44 +10778,78 @@ static void f_jobresize(typval_T *argvars, typval_T *rettv)
-    return;
+  list_T *argl = cmd_tv->vval.v_list;
+  int argc = argl->lv_len;
+  if (!argc) {
```
See above comment.

splinterofchaos replied:
:+1:

splinterofchaos commented:
Rebased and made the code a little more consistent. Thanks @Pyrohh.

Pyrohh commented:
Thanks @splinterofchaos. Given @tarruda and @bfredl both said LGTM already, I'd imagine this can be merged (assuming you think so as well).

justinmk commented on runtime/doc/eval.txt:
```diff
@@ -6299,15 +6301,19 @@ tempname()					*tempname()* *temp-file-name*
+		Spawns {cmd} in a new pseudo-terminal session connected
+		to the current buffer.  {cmd} is the same as the one passed to
+		|jobstart()|.  This function fails if the current buffer is
+		modified (all buffer contents are destroyed).
```
This comment is bizarre. Are the buffer contents replaced, or in fact a new buffer is created, which *may* causes loss of unsaved changes in the case of `set nohidden`?

justinmk replied:
Oh, actually it's the function behavior that's bizarre. Why do we replace the current buffer with a `term:` buffer instead of creating a *new* buffer?

splinterofchaos replied:
At least the first part of that relates #2368. `f_termopen()` has
```c
  if (curbuf->b_changed) {
    EMSG(_("Can only call this function in an unmodified buffer"));
    return;
  }
```
to protect against destroying modified buffers, but doesn't hide it. Also, some reading of code using `b_changed` suggested to me that this check might be insufficient. This causes #2408 because of a static variable in `syntax.c` not getting updated.

`termopen` definitely still needs some work, but I want to save that for another PR so I can get this merged.

justinmk commented on runtime/doc/eval.txt:
```diff
@@ -6299,15 +6301,19 @@ tempname()					*tempname()* *temp-file-name*
+
+		The {opts} dict is similar to the one passed to |jobstart()|,
+		but the `pty`, `width`, `height`, and `TERM` fields are
+		ignored: `height`/`width` are taken from the current window
```
Why are ``pty`, `width`, `height`, and `TERM`` ignored?

splinterofchaos replied:
I can't think of a use case for setting a custom `'height'`, but perhaps `'width'` could be set to the maximum of the buffer's width and the given number so that long commands/output get linewrapped. I seem to remember someone opening an issue about this, but I can't find it.

I'd like to save that for another PR, though. I haven't modified that code.

justinmk commented on runtime/doc/various.txt:
```diff
@@ -224,10 +224,17 @@ g8			Print the hex values of the bytes used in the
 
 						  *:term* *:terminal*
 :term[inal][!] {cmd}	Spawns {command} using the current value of 'shell'
```
This should be `{cmd}`

splinterofchaos replied:
:+1:

justinmk commented on runtime/doc/various.txt:
```diff
@@ -224,10 +224,17 @@ g8			Print the hex values of the bytes used in the
+			                    \{'name':'{cmd}'})
+			      :startinsert
 <
+			If no {command} is given, 'shellcmdflag' will not be
```
{command} => {cmd}

splinterofchaos replied:
This happens when `{cmd}` is not specified, so it's equivalent to
```vim
enew
call termopen(&shell, {'name' : &shell)
startinsert
```
but after looking `ex_terminal` over, it doesn't look like I correctly do that.  Now fixed.

splinterofchaos commented:
I'll mark this RDY, but will wait until tomorrow in case anyone has a last minute comment.

justinmk commented on src/nvim/eval.c:
```diff
@@ -10779,44 +10779,78 @@ static void f_jobresize(typval_T *argvars, typval_T *rettv)
+    if (exe) {
+      EMSG2(e_jobexe, exe);
     }
+    return NULL;
```
if `!exe`, we return NULL without an error message.

justinmk commented on src/nvim/ex_docmd.c:
```diff
@@ -9409,9 +9409,20 @@ static void ex_folddo(exarg_T *eap)
-      eap->forceit==TRUE ? "!" : "",
-      strcmp((char *)eap->arg, "") ? (char *)eap->arg : (char *)p_sh);
-  do_cmdline_cmd((uint8_t *)cmd);
+  // We will call `termopen()` will `[&shell]` if given a {cmd}.
```
will => with

splinterofchaos replied:
The brackets denote that we send it as a list, but using `'shell'` instead of `&shell` does seem to make sense.

splinterofchaos commented:
@justinmk, have I satisfied your concerns?

justinmk commented:
LGTM
